src/software/geom/algorithms/end_in_obstacle_sample.cpp

```diff
@@ -0,0 +1,51 @@
+#include <random>
+#include "software/geom/algorithms/end_in_obstacle_sample.h"
+#include "software/geom/point.h"
+#include "software/ai/navigator/obstacle/obstacle.hpp"
+
+
+Point endInObstacleSample(const std::vector<ObstaclePtr> obstacles, const Point &destination, const Rectangle &navigable_area, double multiplier, double range)
+{
+    for (auto const &obstacle : obstacles) {
+        if (obstacle->contains(destination)) {
+            goto get_nearest;
+        }
+    }
+    return destination;
+
+get_nearest:
+    for (auto const &obstacle : obstacles) {
+        if (obstacle->contains(destination)) {
+
+            Point closest_point = obstacle->closestPoint(destination);
+            for (auto const &obstacle : obstacles) {
+                if (obstacle->contains(closest_point)) {
+                    goto do_sampling;
+                }
+            }
+            return closest_point;
+        }
+    }
+
+do_sampling:
+    double rad = 0.15;
+    int steps_per_rad = 6;
+    while (rad <= range) {
+        double increment = 360.0 / steps_per_rad;
+        for (int i = 0; i < steps_per_rad; i++) {
+            Angle angle = Angle::fromDegrees(float(i) * increment);
```

Comment:
We shouldn't do C-style casting. `static_cast<float>(...)` should be used instead (same with `int(...)` below)